### DISCUSSION TITLE
Configuring build directories and copying files from build directories to serverless functions

### DISCUSSION THREAD

**Discussion by azangru** [vercel/vercel#4154] [General] [answered]
Hi,

I am struggling with Now deployment, and am already at my wits’ end.

Here's what I am trying to achieve:

During the build process, I want to build:
- a small SQLite file for my serverless functions
- a bunch of static html files using Eleventy

### Attempt 1

I wanted to put my SQLite file in the `build` directory, and my static files in the `public` directory. Both directories are created at build time in an npm build script:

```
 mkdirp build && mkdirp public && node ./build.js && npm run eleventy-build && cp -r ./_site/* ./public
```

Here's the relevant part of the now.json config:

```
  "functions": {
    "api/article.js": {
      "includeFiles": "build/database.db"
    }
  }
}  
```

However, the public directory never got picked up by Now. The output directory always remained the root of the project — despite the Build settings page saying that the public folder will be used for output if available.

### Attempt 2

Ok, I gave up on the public folder, and switched to Eleventy framework preset in the Build and Development Settings. Now the output directory is `_site`, and it successfully gets picked up by Now; but it seems that my now.json config settings:

```
  "functions": {
    "api/article.js": {
      "includeFiles": "build/database.db"
    }
  }
}  
```

are no longer getting applied, because serverless functions can no longer access the build/database.db file.

Could you please advise how to set up custom build directories so that they get picked up even if they are created only during the build process?

**Comment by styfle** [comment, marked as answer]
Hi @azangru 

Only one Output Directory can be used to serve static content. This is because the output files map to routes in your deployment. You'll need to chose one Output Directory (it sounds like either `build`, `public`, or `_site`).

If your Serverless Functions need to bundle files different files, you can define a [`now-build`](https://vercel.com/docs/runtimes#advanced-usage/advanced-node-js-usage/custom-build-step-for-node-js) script.

Lastly, I want to mention that sqlite is probably not a good choice for Serverless Functions because the file system is stateless and any writes are not persisted. Additionally, as the function scales up, new instances are invoked and thus any data written to one instance will not be available in another instance.

**Reply by azangru** [reply]
Hi @styfle 

I feel that by describing different issues arising from my different attempts to solve the problem I made it sound more confusing. Let me focus on just one approach and the problem associated with it.

- I've selected the Eleventy preset, and am building static assets and putting them in an output directory (let's say, `_site`). This works fine.
- Also, during the build step, I am building several other files (e.g. the SQLite file) and storing them in a temporary directory. I want to bundle these files with the serverless function
- I indicate my intent to bundle the build artifacts with the serverless functions in the `now.json` config file, by pointing at my temporary directory in the `includeFiles` field
- **I expect** the build artifacts to be now available from within the serverless functions
- **Problem:** serverless functions cannot access the build artifacts

Are my expectations wrong? I believe the problem occurs because the folder where I am storing build artifacts for bundling into serverless functions is temporary, i.e. it does not exist when the repository gets pulled by the now builder. However, since this particular problem occurs only when a build preset is selected (without the preset, the problem will manifest differently), I also suspect that the now.json file does not get executed upon when there is a build preset.

My npm script for building the assets to be bundled with serverless functions is named `build`. Will Now behave differently if it is named `now-build`?

As to your point about sqlite, I am only using it for reads of the data generated at the during the build step.

**Reply by azangru** [reply]
@styfle now that I've changed the name of my npm `build` script to `now-build`, everything is working perfectly! Thank you so much for your help.

As a feedback from a user who was reading the docs and using the UI of Now, may I suggest that you make it more explicit in the docs how the `now-build` script is different from `build`? From the interface, I got an impression that `now-build` does the same as `build` (see placeholder text in the Build command field):

![image](https://user-images.githubusercontent.com/6834224/80363217-ca9f1680-887b-11ea-8f71-e2878e8439b4.png)

and after switching to the Eleventy preset, I thought that the `now-build` will not even be recognized (see placeholder text again):

![image](https://user-images.githubusercontent.com/6834224/80363345-02a65980-887c-11ea-9213-bc7246e69de1.png)

**Reply by styfle** [reply]
Indeed it is confusing. This is an advanced feature so its not available in the UI and would be difficult considering each function could have it's own package.json with a `now-build` script.

I'll forward your feedback to the team, thanks!
